src/chapters/08-tools.md

```diff
@@ -0,0 +1,611 @@
+                    output += fmt.Sprintf("  \"result_%d\": %q", i+1, result)
+                    if i < len(results)-1 {
+                        output += ","
+                    }
+                    output += "\n"
+                }
+                output += "}"
+            case "summary":
+                output = fmt.Sprintf("System Information Summary (%d items):\n", len(results))
+                for _, result := range results {
+                    output += result
+                }
+            default: // "text" or any other value
+                for _, result := range results {
+                    output += result
+                }
+            }
+
+            return output, nil
+        })
+}
+```
+
+In the example above, the `systemInfo` tool accepts a complex input structure that allows users to specify:
+
+- `info_types`: Types of information to gather (e.g., environment variables, working directory, hostname, user). It is an array of strings, allowing multiple types to be requested at once.
+- `env_vars`: Specific environment variables to retrieve.
+- `format`: Output format (e.g., JSON, text, summary).
+- `include_empty`: Whether to include empty or unset values.
+- `max_length`: Maximum length for each value (0 means no limit).
+
+Example usage scenarios:
+
+```bash
+User: "Show me system information in JSON format"
+AI calls: systemInfo({
+  "info_types": ["env", "workdir", "hostname", "user"],
+  "format": "json",
+  "include_empty": false
+})
+```
+
+Here is another example:
+
+```bash
+User: "Get the HOME and PATH environment variables, limit each to 50 characters"
+AI calls: systemInfo({
+  "info_types": ["env"],
+  "env_vars": ["HOME", "PATH"],
+  "format": "text",
+  "max_length": 50,
+  "include_empty": false
+})
+```
+
+This pattern is particularly useful when you need tools that can adapt their behavior based on detailed user requirements, providing flexibility while maintaining type safety and clear documentation through JSON schema descriptions.
+
+Here is an example of how the tool can be used in the UI, in the trace you can see the tool being called with complex parameters being filled in by the AI model thanks to the JSON schema description, the prompt was `give me information about the current user in json format`:
+
+![](../images/chapter-08/complex-input.png)
+
+
+## Running the Example
+
+To run the Chapter 08 example:
+
+1. **Set up environment**:
+
+   ```bash
+   export OPENAI_API_KEY="your-api-key"
+   ```
+
+2. **Run the application**:
+
+   ```bash
+   cd src/examples/chapter-08
+   go run main.go
+   ```
+
+3. **Test with curl**:
+
+   ```bash
+   curl -X POST http://localhost:9090/operatingSystemFlow \
+     -H "Content-Type: application/json" \
+     -d '{"data": "What time is it right now?"}'
+   ```
+
+4. **Invoke with Genkit CLI**
+
+    ```bash
+    cd src/examples/chapter-08
+    genkit flow:run operatingSystemFlow '"List all directories in folder /"'
+    ```
```

Comment:
The `genkit flow:run` command requires `genkit start` to be running. 
If use `go run main.go`, you'll encounter this Error: No runtimes found. Make sure your app is running using 'genkit start -- ...'.
Therefore, it would be helpful to mention that you need to start the server with `genkit start -- go run  main.go` before executing `genkit flow:run operatingSystemFlow`.
(Note that the curl command works fine with just go run main.go.)

Reply:
that is absolutely true